runtime/io/common/src/aws/smithy/kotlin/runtime/io/SdkManaged.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.io
+
+import aws.smithy.kotlin.runtime.util.InternalApi
+import kotlinx.atomicfu.locks.SynchronizedObject
+import kotlinx.atomicfu.locks.synchronized
+
+/**
+ * Interface that the lifecycle of some resource is managed by the SDK at runtime.
+ */
+@InternalApi
+public interface SdkManaged {
+    /**
+     * Invoked by a caller to declare usership of the resource.
+     */
+    public fun share()
+
+    /**
+     * Invoked when a caller releases the resource. Returns a boolean indicating whether the resource has been released
```

Comment:
**nit**: The semantics here are just `share` and `unshare`, the concept of what happens to the resource when it is unshared is up to the implementaiton. Maybe remove the language here referring to "release", etc.